### PR TITLE
Added TenX Reference Command call

### PR DESCRIPTION
**Pull request by saurabsa**
- Kubernetes does not allow '-' in their ENV variable names.
- Refactor

### CODE REVIEW

**Comment by xinyanmsft** [comment]
Is there a PR on the c# code that implements ‘reference add’ command?
Thanks!
Xin

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py [hunk 1]
```diff
@@ -342,6 +342,7 @@ def add_reference(target_group, target_name, reference_name):
     """
     service_name = _get_service_name()
     instance, client = references.get_reference_type(target_group, target_name)
+    utils.write("Adding reference ...")
```
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: This is not in the walkthrough/demo, so let's remove it.

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py [hunk 2]
```diff
@@ -8,23 +8,23 @@ def get_environment_var_name(service_name, reference_name):
     """
     Gets the environment variable name for a reference
     """
-    return '{}_{}_url'.format(service_name, reference_name).upper()
+    return '{}_{}_url'.format(service_name.replace('-', ''), reference_name.replace('-', '')).upper()
```
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: why this change?

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py [hunk 3]
```diff
@@ -8,23 +8,23 @@ def get_environment_var_name(service_name, reference_name):
...
 
 
 def create_connection_string(service_name, reference_name, connection_string):
     """
     Creates a secret on Kubernetes that stores
     the connection string and is labeled with run=service_name
     """
-    secret_name = '{}-{}'.format(service_name, reference_name)
+    secret_name = '{}-{}'.format(service_name.replace('-', ''), reference_name.replace('-', ''))
```
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: ?

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py [hunk 4]
```diff
@@ -8,23 +8,23 @@ def get_environment_var_name(service_name, reference_name):
...
     encoded_connection_string = base64.b64encode(
         bytes(connection_string, 'utf-8'))
     environment_variable_name = get_environment_var_name(
         service_name, reference_name)
 
     # Create the secret
     command = 'kubectl create secret generic {} --from-literal={}={}'.format(
-        secret_name, environment_variable_name, encoded_connection_string)
+        secret_name.lower(), environment_variable_name.lower(), encoded_connection_string)
```
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: if this needs to be .lower(), the it should be done in ```get_environment_var_name``` andwhere we create the ```secret_name```

**Comment by saurabsa** [comment]
Yes I will send it today. I won't merge this one this InnerLoop PR is also approved.

Get Outlook for Android<https://aka.ms/ghei36>


________________________________

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py (hunk 2 above)
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: Kubernetes doesnot allow ENV name with '-'

---
In reply to: [109705032](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109705032) [](ancestors = 109705032)

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py (hunk 3 above)
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: Trying to keep the same name for secret, key and ENV should that I have to just share less strings to the tenx server for deployment.

---
In reply to: [109705054](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109705054) [](ancestors = 109705054)

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py (hunk 4 above)
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: I just didn't want to get lower case ENV variable when you would call the get method.

---
In reply to: [109705224](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109705224) [](ancestors = 109705224)

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py (hunk 3 above)
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: We shouldn't be modifying the reference name as it was provided by the user. If it contains unsupported characters (which it shouldn't as we are not using it for any kubernetes resource), we should fail even before we try to create the connection string.

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py (hunk 4 above)
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: Secret name should be lower case and checked for unsupported characters, however, we should leave the environment variable name upper case

Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/references.py]: - Yes secret name should be lower case, that's why I was trynig to put lower() here in case user has provided a non all lower reference name.
 - ENV name has to be upper, that's why I didnt make the change in the get method. But this is the key name where you are using the same ENV name, so that one has to be lower.

---
In reply to: [109718286](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109718286) [](ancestors = 109718286)

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py [hunk 5]
```diff
@@ -359,13 +362,18 @@ def add_reference(target_group, target_name, reference_name):
...
 
-    utils.writeline("\nAdded reference '{}'".format(reference_name))
+    utils.writeline("Added reference '{}'".format(reference_name))
     utils.writeline('Environment variables: {}'.format(
         references.get_environment_var_name(service_name, reference_name)))
     
     _service_add_reference(reference_name, instance.type)
 
 
+def _validate_reference_name(reference_name):
```
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: add cpmments

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py [hunk 6]
```diff
@@ -266,6 +266,14 @@ def add_reference(self, service_name, reference_name, reference_type):
         self.settings['references'] = references
         self._save_changes()
 
+    def reset_references(self):
```
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py]: Why do you need this?

Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py]: I used the command: azp create -f 
and it created a project for me with the same reference list as it had for the old project. I assume we don't clear the reference list when creating a new project unless there is no existing project.

---
In reply to: [109784646](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109784646) [](ancestors = 109784646)

Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py]: isn't calling azp create the second time just sleeping for 5 seconds? I thought that's what the plan is, so we shouldn't be 'clearing' the projectReferences file right now.

Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py]: Yes but with the -f flag, it force creates the project.

---
In reply to: [109793069](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109793069) [](ancestors = 109793069)

Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py]: when you do -f it still creates you a complete new set of infrastructure - e.g. if we didn't use random names for resource group or resources you'd get the same error for something else (e.g. storage account exists, resource group exists, etc.). I would remove the -f as we dont actually know what does it mean to force the project creation - is it re-creating everything from scratch? or just recreating part of resources (e.g. keep the infrastructure, configure the cluster again?).

Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/settings.py]: I agree. I just added it for faster tests. I will remove it.

---
In reply to: [109794766](https://github.com/saurabsa/azure-cli/pull/43#discussion_r109794766) [](ancestors = 109794766)
